### ISSUE TITLE
compilation_db does not seperate by build variant when using SConscript

### ISSUE THREAD

**Issue by andreaskremser** [SCons/scons#3742]
**Describe the bug**
im trying to use the compilation_db Tool for several variants with multiple SConscript('SConscript', variant_dir='build/' + variant, duplicate=0) invocations (in a for each loop). With the current scons from git each compile_commands.json includes all the build commands for every variant. (i.e. build/win64/compile_commands.json will be the same as build/linux/compile_commands.json, and both will contain all win64+all linux compile commands).

The result will be multiple commands per source file, each pointing to a different build variant. 
e.g. for one file with 3 variants, each compile_commands.json looks like that:

>   cat build/linux/compile_commands.json
```
[
    {
        "command": "gcc -o build/linux/main.o -c main.c",
        "directory": "/home/akr/projects/experiments/scons_variant",
        "file": "main.c",
        "output": "build/linux/main.o"
    },
    {
        "command": "gcc -o build/qnx/main.o -c main.c",
        "directory": "/home/akr/projects/experiments/scons_variant",
        "file": "main.c",
        "output": "build/qnx/main.o"
    },
    {
        "command": "gcc -o build/simul/main.o -c main.c",
        "directory": "/home/akr/projects/experiments/scons_variant",
        "file": "main.c",
        "output": "build/simul/main.o"
    }
]
```
**Required information**
* Link to SCons Users thread discussing your issue.
https://discord.com/channels/571796279483564041/571796280146133047/731887216212574298
* Version of SCons
4.0 & latest git (cd722bdc5f6b1163d56246ee0afc63c28ecc138e)
* Version of Python
3.8.3
* Which python distribution if applicable (python.org, cygwin, anaconda, macports, brew,etc)
vanilla? (fedora, debian, guix)
* How you installed SCons
git clone
* What Platform are you on? (Linux/Windows and which version)
fedora 32, debian testing, guix latest
* How to reproduce your issue?  Please include a small self contained reproducer. Likely a SConstruct should do for most issues.
a small reproducible example is attached. Just run scons on it.
* How you invoke scons (The command line you're using "scons --flags some_arguments")
"scons"
[scons_variant_compiledb.zip](https://github.com/SCons/scons/files/4908933/scons_variant_compiledb.zip)

**Comment by bdbaddog** [comment]
Hmm.  Are you setting duplicate=0 or duplicate=1 for your variant dirs?

**Comment by andreaskremser** [comment]
(Note: i added more information to the SConscript call in the description to make it more clear)
In the provided example i'm using duplicate=0. With duplicate=1 (SConscript('SConscript', variant_dir='build/' + variant, duplicate=1)) i still get multiple entries per file but with compile commands that refer to the individual copies.

> cat build/linux/compile_commands.json 
```
[
    {
        "command": "gcc -o build/linux/main.o -c build/linux/main.c",
        "directory": "/home/akr/projects/experiments/scons_variant",
        "file": "main.c",
        "output": "build/linux/main.o"
    },
    {
        "command": "gcc -o build/qnx/main.o -c build/qnx/main.c",
        "directory": "/home/akr/projects/experiments/scons_varian",
        "file": "main.c",
        "output": "build/qnx/main.o"
    },
    {
        "command": "gcc -o build/simul/main.o -c build/simul/main.c",
        "directory": "/home/akr/projects/experiments/scons_variant",
        "file": "main.c",
        "output": "build/simul/main.o"
    }
]
```

**Comment by bdbaddog** [comment]
So perhaps a way to do this would provide a way to provide CompilationDatabase with a set of target filters?

**Comment by bdbaddog** [comment]
@acmorrow - any thoughts on this?

**Comment by bdbaddog** [comment]
Marking this an enhancement request.
User on exjam on discord server said this doesn't break any tooling, but some (many/all) consumers may only look at the first entry for a source file and not look at others.  They may take a shot at creating a PR for this.

```
for example, if you used compile_commands.json in an IDE like VSCode to provide the information required for "intellisense", then it makes sense to have a per-config .json. VSCode did handle having multiple builds in a single compile_commands.json but it was probably just using the first one it found rather than exploding. if you had them in different files you could switch between the configurations in the IDE
bdbaddogToday at 8:04 AM
IC. so really it's not causing a real problem, just a preference?
exjamToday at 8:04 AM
yeah i guess
i've only tried it as input for clang-tidy, and in VSCode to power intellisense and both worked without breaking
```

**Comment by andreaskremser** [comment]
Hi thanks again!

Thats right most of the tools just take the first entry and then seem to work. But that can result in wildly inconsistent results. E.g. for windows/linux specific builds, ifdefs, code passed as define, different files used only for different builds, etc. So in a certain sense it can break tooling.

In the meantime i tried this [scons_compiledb](https://github.com/pinetr2e/scons-compiledb) plugin as a workaround.
It looks similar to the one in SCons mainline but does not seem to use global state, and thus can be used to create different compiledb's for each variant. There might be some lines of code worth absorbing into the mainline plugin :)

Its available via pip

`pip install scons-compiledb`


It can be used to solve the issue like that:

```
import scons_compiledb

linux_config = scons_compiledb.Config(db="build/linux/compile_commands.json)
windows_config = scons_compiledb.Config(db="build/lwindows/compile_commands.json)

scons_compiledb.enable(linux_env, linux_config)
scons_compiledb.enable(windows_env, windows_config)

linux_env.Compiledb()
windows_env.Compiledb()
```

**Comment by bdbaddog** [comment]
@andreaskremser - interesting. That plugin says bases on mongodb, our functionality is as well. I'll have to take a look at it sometime.

**Comment by bdbaddog** [comment]
Fixed by #3818
